File: Dynamic-Programming/354-H-Russian-Doll-Envelopes/russian_doll.py
class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        envelopes.sort(key=lambda x: (x[0], -x[1]))
        DP = []

        for w, h in envelopes:
            i = bisect_left(DP, h)
            if i==len(DP):
                DP.append(h)
            else:
                DP[i] = h
        return len(DP)

File: Dynamic-Programming/354-H-Russian-Doll-Envelopes/russian_doll.cpp
class Solution {
public:
    int maxEnvelopes(vector<vector<int>>& envelopes) {
        if (envelopes.empty()) 
            return 0;

        sort(envelopes.begin(), envelopes.end(), [](const vector<int>& a, const vector<int>& b) {
            if (a[0] == b[0]) { return a[1] > b[1];}
            return a[0] < b[0];
        });

        vector<int> tails;

        for (const auto& env : envelopes) {
            int h = env[1];
            
            auto it = lower_bound(tails.begin(), tails.end(), h);
            
            if (it == tails.end()) {
                tails.push_back(h);
            } else {
                *it = h;
            }
        }
        return tails.size();
    }
};